### DISCUSSION TITLE
How the dimintions of x0 is specified?

### DISCUSSION THREAD

**Discussion by wissamkafa** [do-mpc/do-mpc#379] [Q&A] [answered]
I am trying to apply the model used in "rotating_oscillating_masses_mhe_mpc" to another mechanical system. it works with mhe but when i tried to implement it with state feedback only, I always get the error shown below.
the initial values of x0 is a (6,1) vector as the number of "_x" variables in the model, but when the complier reaech this code
for i in range(200):
    u0 = mpc.make_step(x0)
    x0 = simulator.make_step(u0)
it works for the first step as the initial dimintions of x0 are correct, but in the second iteration it seems that the 'x0' dimintions change to (4,1) (as the debugger shows).
I could not figure out why and which function changes it, or eve why it changes (as the model contain 4 states variables "_x" and another two for measurment "_x") (so, it should be 6)
#============Error
Traceback (most recent call last):
  File "/home/user/PhD/.../main.py", line 195, in <module>
    u0 = mpc.make_step(x0)
  File "/home/user/anaconda3/lib/python3.10/site-packages/do_mpc/controller/_mpc.py", line 1024, in make_step
    assert n_val == self.model.n_x, 'Wrong input with shape {}. Expected vector with {} elements'.format(n_val, self.model.n_x)
AssertionError: Wrong input with shape 4. Expected vector with 6 elements
#===========
so, please can you help me with this.

**Comment by 4flixt** [comment, marked as answer]
Hi, 
the simulator does not return the state but the measurements of the system. If they are not set in the model, it defaults to full state-feedback (and does indeed return the state).
The model from the example ``rotating_oscillating_masses_mhe_mpc`` does define measurements, however, as it is used to demonstrate the estimator. 
A simple fix is to just to remove all lines that look like this:
```python
model.set_meas(...)
```
from the model.

**Reply by wissamkafa** [reply]
Thanks for your help (removing the measuremnt states solved the problem).

I already turned around the problem by adding two extra state measuremnt (dphi_meas = model.set_meas('dphi_meas', dphi))

The code works, but actualy I find some difficulties to connect things together it needs careful reading of the document but sometimes things become ambigous. 

But i will keep going as I am planning to use the library in my research.

one more question please:
if we define a vector variable like in this line, 
dphi = model.set_variable(var_type='_x', var_name='dphi', shape=(2,1))

how could we deal with its elements individually when adding and plotting lines
sim_plot.add_line('_x', 'dphi', ax[3])
the variable name should be string and when using l=dphi[0] for examle, it does not work, is there any methode to overcome it?
